File: dart-to-javascript-interop-node/udp/udp.js
var datagram = require('dgram');

class Udp {
  socket = datagram.createSocket('udp4');

  constructor() {
    console.log("I am javascript");
  }

  listen(port, onListening, onMessage, onError) {
    this.socket.bind(port);
    
    this.socket.on('message', function(msg, rinfo) {
      onMessage(msg, rinfo);
    });
    
    this.socket.on('error', function (err) {
      onError(err);
      socket.close();
    });

    this.socket.on('listening', () => {
      var address = this.socket.address();
      onListening(address.address, address.port);
    });
  }

  send(message, address, port) {
    this.socket.send(message, 0, message.length, port, address);
  }
  
  close() {
    this.server.close();
  }
}

module.exports = Udp;

File: dart-to-javascript-interop-node/udp/udp.dart
@JS()
library udp;

import 'dart:typed_data';

import 'package:js/js.dart';

@JS()
class Udp {
  external factory Udp();
  external void listen(int port, Function onListening, Function onMessage, Function onError);
  external void send(Uint8List message, String address, int port);
  external void close();
}

@JS()
@anonymous
class RInfo {
  external String get address;
  external String get family;
  external int get port;
  external int get size;

  external factory RInfo({
    String address,
    String family,
    int port,
    int size,
  });
}
